text/0000-semicolon-items.md

```diff
@@ -208,6 +258,90 @@ be emitted at once, and not one by one:
 Thereby, the fewer edit-compile cycles will be required and thus the flow
 is disturbed less, but still to a degree.
 
+## Add an `--fpermissive` mode to the compiler
+
+Similar to how `gcc` provides the flag `-fpermissive` which treats *all*
+"sufficiently recoverable" errors as warnings and continues compilation until
+final lib/exe is produced, a Rust compiler could also provide a similar flag.
+This could also be done by writing:
+
+```rust
+#[allow(non_fatal_errors)]
+```
+in the source code.
+
+However, such flags have a few drawbacks:
+* Flags are less likely to be seen as a requirement of the language
+  specification itself. As a consequence, future alternative implementations
+  of Rust will have less of an imperative to support such a flag,
+  making it more likely to develop language dialects, which we've heretofore
+  avoided.
+
+* Not all recoverable warnings are equal, and some may be more indicative of a
+  logic error than a redundant semicolon is. Being able to lint against `;`
+  specifically and configure that is a more limited change.
+
+* The hassle of remembering to use `-fpermissive` or something like it outdoes
+  any gains made by it.
+
+## Do not lint
+
+A more permissive option is to not provide any lint, or make it an
+allow-by-default lint. We reject this because:
+
+1. The following unlikely, but problematic, mistake becomes difficult to spot:
+
+   ```rust
+   #[my_attr]; // <-- hard to spot!
+   fn foo() { ... }
+   ```
+
+   Here, the attribute applies to the semicolon, not the `fn`, but that is
+   not easy to spot at all.
+
+2. We lint today against superfluous punctuation in the form of parenthesis in
+   type and expression contexts. Linting against extraneous semicolons is
+   therefore more consistent.
+
+## Interpret `;` as whitespace at the top level
+
+This solution would simply ignore `;` at the module level or inside `impl`s.
+This will however have too far reaching negative implications on `;` as a
+separator in macros which it is frequently used as.
+
+## Allow an optional `;` after all items
+
+A different technical solution that nonetheless achieves the set out motivations
+of this RFC is to allow syntax like: `struct F {};` specifically. This is then
+parsed as a single item rather than `struct F {}` and `;` as two separate items.
+This is less consistent with the treatment of `;` inside `fn` bodies as well as
+in other languages. 
+
+In addition, the following macro:
+
+```rust
+macro_rules! m {
+    ($($item: item);*) => {}
+}
+```
+
+which today accepts the invocation:
+
+```rust
+m!(fn f() {}; fn g() {});
+```
+
+would no longer do so if `fn f() {};` is interpreted as one `item`,
+because `m!` is expecting a semicolon `;` after an `item`. That semicolon
+has now been gobbled, and so the invocation will result in an error.
+
+## Allow extraneous `;` wherever `item`s are allowed
```

Comment:
It seems like this proposal is closest to what happens with statements. `;`s are _not_ statements, they are just gobbled up when found in a block context ([source, I think](https://github.com/rust-lang/rust/blob/fff1abadd7a4ec861ca4b9c77035379578ef033d/src/libsyntax/parse/parser.rs#L4662))

You can verify this, [`$x:stmt` doesn't match semis](http://play.rust-lang.org/?gist=b3e64ec7d0a39a66cf3bc5c2200211f9&version=stable&mode=debug).

Given that a lot of this RFC is motivated by similarity with statements, perhaps we should go this route?

This will likely have the least impact on macros simply because this doesn't affect macro invocation parsing at all, much like the existing support for `;` in blocks doesn't affect how statements work in a macro context.

Reply:
cc @ExpHP -- what do you think about this technical solution?

This does seem more consistent. It also conveniently sidesteps the issue of doc-comments and attributes on `;` as an item which @scottmcm brought up.

The main drawback is that given a macro in the style `macro_rules! mac { ($($item:item)*) => { .. } )`, it won't accept `mac!(struct F {};)`. But that seems like something that doesn't occur nearly often enough.

At the very least, this sounds like a good conservative starting point.

Reply:
Yeah, also, that's already the case and that's how statements work in macros, again.

Reply:
Fundamentally `;` is _not_ an item, and treating it as such to make parsing nicer doesn't seem like a good fix overall especially when in Rust this is way more important because of our semantic macros. Because you're trading off some parsing annoyances for compiler complexity in how attributes, proc macros, macros, and doc comments work, and introducing a bunch of new edge cases.